[quant][graphmode] FP16 quant support - Operator Fusion

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40710 [quant][graphmode] FP16 quant support - Operator Fusion**
* #40709 [quant][graphmode] FP16 quant support - Insert cast operators
* #40708 [quant][graphmode] Add FP16 quant support - Insert Noop Observers

Summary:

Test Plan:
python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22335975](https://our.internmc.facebook.com/intern/diff/D22335975)

supriyar committed:
[quant][graphmode] FP16 quant support - Operator Fusion
Summary:

Test Plan:
python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graphmode] FP16 quant support - Operator Fusion"


Summary:

Test Plan:
python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on test/quantization/test_quantize_jit.py:
```diff
@@ -2982,3 +3002,17 @@ def test_single_linear_dynamic(self):
+
+        for model in [model_traced, model_script]:
+            model_quantized = quantize_dynamic_jit(model, qconfig_dict, debug=False)
+            # TODO check model with debug=True matches quantized model result
```
you can also check numerics with `checkGraphModeOp`

supriyar replied:
Yeah, right now looks like debug option set to `True` isn't producing identical numerics. Once I fix/figure that out I'll use `checkGraphModeOp`

jerryzh168 commented on torch/csrc/jit/passes/quantization/finalize.cpp:
```diff
@@ -22,10 +22,30 @@ graph(%a_dequant, %w_quant, %b):
+    const auto& pattern = entry[0];
+    const auto& replacement = entry[1];
+    rewriter.RegisterRewritePattern(pattern, replacement);
+    rewriter.runOnGraph(graph);
```
don't you need to check the values of the arguments in the pattern: %dtype_fp16, %dtype_fp32, %default_param, %non_blocking?

jerryzh168 replied:
Example: https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/passes/quantization/helper.cpp.html#_ZN5torch3jit21aten_add_alpha_is_oneERKNS0_5MatchERKSt13unordered_mapINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEPNS0_5ValueESt4hashISA_ESt8e5539190

supriyar replied:
Yes, I was looking for this. Thanks!

jerryzh168 commented on torch/csrc/jit/passes/quantization/quantization_patterns.h:
```diff
@@ -1007,8 +1007,25 @@ graph(%packed_params, %a, %reduce_range, %a_dtype):
+
   return {
       {"quantized::linear_dynamic", linear_dynamic, quantized_linear_dynamic},
+      {"quantized::linear_dynamic_fp16",
```
same here

jerryzh168 replied:
maybe it's fine to not check here since it's checked before

supriyar replied:
Done

supriyar committed:
Update on "[quant][graphmode] FP16 quant support - Operator Fusion"


Summary:

Test Plan:
python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization/finalize.cpp:
```diff
@@ -7,6 +7,12 @@
 namespace torch {
 namespace jit {
 
+struct PatternReplaceInfo {
```
we have this in quantization_patterns.h

jerryzh168 replied:
might make sense to move the patterns there as well

supriyar replied:
I can do this in a follow-up PR and move all patterns in this file there.

jerryzh168 commented on torch/csrc/jit/passes/quantization/helper.cpp:
```diff
@@ -668,5 +668,17 @@ bool is_batchnorm3d_module(
+  const auto& match_vmap = match.values_map;
+  auto fp16_type = toIValue(match_vmap.at(vmap.at("dtype_fp16")));
+  auto fp32_type = toIValue(match_vmap.at(vmap.at("dtype_fp32")));
+  auto default_param = toIValue(match_vmap.at(vmap.at("default_param")));
```
I think it's better to split this check to a separate one, might be useful for other checks as well

jerryzh168 replied:
also, you can name default_param to `%false` so that this check can be used by all values that needs to be false

jerryzh168 replied:
I took a look at https://pytorch.org/docs/stable/tensors.html#torch.Tensor.half, I think it might make more sense to have `half` and `float` here

jerryzh168 replied:
non_blocking sounds irrelevant here.

jerryzh168 commented on torch/csrc/jit/passes/quantization/finalize.cpp:
```diff
@@ -22,10 +28,30 @@ graph(%a_dequant, %w_quant, %b):
+      {linear_with_quant, linear_with_quant_prepack},
+      {linear_fp16_with_cast,
+       linear_fp16_with_prepack,
+       {is_fp16_fp32_cast_op}}};
```
non_blocking is still not checked

supriyar replied:
I think that param doesn't have to be a specific value, so I omitted that check

jerryzh168 replied:
did we insert the non_blocking = false code? I think it is better to check this as well, I'm not sure how non_blocking = true can be simulated with the swapped pattern

supriyar replied:
Actually non-blocking and copy are checked with the `false` check. The missing one is `memory_format` which should be None - I'll check for that.

jerryzh168 commented on torch/csrc/jit/passes/quantization/helper.cpp:
```diff
@@ -668,5 +668,17 @@ bool is_batchnorm3d_module(
+    const Match& match,
+    const std::unordered_map<std::string, Value*>& vmap) {
+  const auto& match_vmap = match.values_map;
+  auto fp16_type = toIValue(match_vmap.at(vmap.at("dtype_fp16")));
+  auto fp32_type = toIValue(match_vmap.at(vmap.at("dtype_fp32")));
```
also would be good to split these as well, so the filters can be more composable

jerryzh168 replied:
you can add a function similar to `is_int_constant` for scalar type

supriyar committed:
Update on "[quant][graphmode] FP16 quant support - Operator Fusion"


Summary:

Test Plan:
python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]
